packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -538,162 +539,118 @@ impl<'m> RunnerImpl<'m> {
         })
     }
 
-    unsafe fn new_buffer(&self, ptr: *const u8, len: usize, _capacity: usize) -> Buffer {
-        let s = std::slice::from_raw_parts(ptr, len);
-        s.into()
-    }
-
     /// TODO: DOC, flushing from a single column
     fn array_data_from_js(
         &mut self,
         mv8: &'m MiniV8,
         data: &mv8::Value<'m>,
-        dt: &DataType,
+        field: &Field,
         len: Option<usize>,
     ) -> Result<ArrayData> {
         // `data` must not be dropped until flush is over, because
         // pointers returned from FFI point inside `data`'s ArrayBuffers' memory.
-        let obj = data
-            .as_object()
-            .ok_or_else(|| Error::Embedded("Flush data not object".into()))?;
-        let child_data: mv8::Array<'_> = obj.get("child_data")?;
+        let obj = data.as_object().ok_or_else(|| {
+            Error::Embedded(format!("Flush data not object for field {:?}", field))
+        })?;
 
         // `data_node_from_js` isn't recursive -- doesn't convert children.
         let data: mv8::DataFfi = mv8.data_node_from_js(data);
 
-        let n_children = child_data.len();
-        let child_data: Vec<ArrayData> = match dt.clone() {
-            DataType::List(field) => {
-                let child: mv8::Value<'_> = child_data.get(0)?;
-                Ok(vec![self.array_data_from_js(
+        // This target length is used because the JS repr does not mirror buffer building as Rust
+        // Arrow and pyarrow.
+        let target_len = len.unwrap_or(data.len);
+
+        // TODO: Extra copies when generating buffers, because JS Arrow doesn't align things
+        //   properly. (Due to which buffer capacities are currently unused.)
+        let mut builder = ArrayData::builder(field.data_type().clone());
+
+        match field.data_type() {
+            DataType::Boolean => unsafe {
```

Comment:
I'm guessing the unsafe is to do with the `slice::from_raw_parts`? Perhaps we could add a safety comment above the match or something to describe why it's safe. 

Out of curiosity why have you also put the builder code within the unsafe, are their operations unsafe if the values array is borked?